src/api.jl

```diff
@@ -0,0 +1,10962 @@
+XPRSsetcheckedmode.
+
+*Synopsis*
+
+int XPRS_CC XPRSgetcheckedmode(int* r_checked_mode);
+
+*Arguments*
+
+`r_checked_mode`: Variable that is set to 0 if checking and validation of Optimizer function calls is disabled for the current process, non-zero otherwise.
+
+*Example*
+
+
+
+*Related*
+
+XPRSsetcheckedmode.
+"""
+function getcheckedmode(r_checked_mode)
+    Lib.XPRSgetcheckedmode(r_checked_mode)
+end
+
+function license(lic, path)
+    Lib.XPRSlicense(lic, path)
+end
+
+function beginlicensing(r_dontAlreadyHaveLicense)
+    Lib.XPRSbeginlicensing(r_dontAlreadyHaveLicense)
+end
+
+function endlicensing()
+    Lib.XPRSendlicensing()
+end
+
+"""
+
+    getlicerrmsg()
+
+*Purpose*
+
+Retrieves an error message describing the last licensing error, if any occurred.
+
+*Synopsis*
+
+int XPRS_CC XPRSgetlicerrmsg(char *buffer, int length);
+
+*Arguments*
+
+`buffer`: Buffer long enough to hold the error message (plus a null terminator).
+`length`: Length of the buffer. This should be 512 or more since messages can be quite long.
+
+*Example*
+
+The following calls
+XPRSgetlicerrmsg to find out why
+XPRSinit failed:
+
+*Related*
+
+XPRSinit.
+"""
+function getlicerrmsg()
+    msg = Cstring(pointer(Array{Cchar}(undef, 1024*8)))
+    len = length(msg)
+    Lib.XPRSgetlicerrmsg(msg, len)
+    return unsafe_string(msg)
+end
+
+"""
+
+    setlogfile(prob::XpressProblem, logname)
+
+*Purpose*
+
+This directs all Optimizer output to a
+log file.
+
+*Synopsis*
+
+int XPRS_CC XPRSsetlogfile(XPRSprob prob, const char *filename);
+
+*Arguments*
+
+`prob`: The current problem.
+`filename`: A string of up to
+`MAXPROBNAMELENGTH characters containing the file name to which all logging output should be written. If set to`: NULL, redirection of the output will stop and all screen output will be turned back on (except for DLL users where screen output is always turned off).
+
+*Example*
+
+The following directs output to the file
+logfile.log:
+
+*Related*
+
+XPRSaddcbmessage.
+"""
+function setlogfile(prob::XpressProblem, logname)
+    Lib.XPRSsetlogfile(prob, logname)
+end
+
+"""
+
+    setintcontrol(prob::XpressProblem, _index, _ivalue)
+
+*Purpose*
+
+Sets the value of a given integer
+control parameter.
+
+*Synopsis*
+
+int XPRS_CC XPRSsetintcontrol(XPRSprob prob, int ipar, int isval);
+
+*Arguments*
+
+`prob`: The current problem.
+`ipar`: Control parameter whose value is to be set. A full list of all controls may be found in
+`Control Parameters, or from the list in the`: xprs.h header file.
+`isval`: Value to which the control parameter is to be set.
+
+*Example*
+
+The following sets the control
+PRESOLVE to
+0, turning off the presolve facility prior to optimization:
+
+*Related*
+
+XPRSgetintcontrol,
+XPRSsetdblcontrol,
+XPRSsetstrcontrol.
+"""
+function setintcontrol(prob::XpressProblem, _index, _ivalue)
```

Comment:
Clang.jl has an option to enforce type?
thats something I have been thinking about for a while... because these C functions can easily (and silently) break if types are wrong

Reply:
Clang.jl does have an [option](https://github.com/JuliaInterop/Clang.jl/blob/7823753f1c5c77a4678f0f33f92b03520962d044/src/wrap_c.jl#L48) but it's [not exposed in their wrapper convenience functions](https://github.com/JuliaInterop/Clang.jl/blob/03ddf229023fd17ea96c61597bfbea73581a73df/src/compat.jl#L33-L47). 

I ran it on this project to show as an example what it does. This is just a sample of the changes to `src/lib.jl`.

```
diff --git a/src/lib.jl b/src/lib.jl
index b7fb2cf..0704c8e 100644
--- a/src/lib.jl
+++ b/src/lib.jl
@@ -2,15 +2,15 @@
 # Automatically generated using Clang.jl


-function XPRScopycallbacks(dest, src)
+function XPRScopycallbacks(dest::XPRSprob, src::XPRSprob)
     ccall((:XPRScopycallbacks, libxprs), Cint, (XPRSprob, XPRSprob), dest, src)
 end

-function XPRScopycontrols(dest, src)
+function XPRScopycontrols(dest::XPRSprob, src::XPRSprob)
     ccall((:XPRScopycontrols, libxprs), Cint, (XPRSprob, XPRSprob), dest, src)
 end

-function XPRScopyprob(dest, src, probname)
+function XPRScopyprob(dest::XPRSprob, src::XPRSprob, probname)
     ccall((:XPRScopyprob, libxprs), Cint, (XPRSprob, XPRSprob, Cstring), dest, src, probname)
 end

@@ -18,7 +18,7 @@ function XPRScreateprob(_probholder)
     ccall((:XPRScreateprob, libxprs), Cint, (Ptr{XPRSprob},), _probholder)
 end

-function XPRSdestroyprob(_prob)
+function XPRSdestroyprob(_prob::XPRSprob)
     ccall((:XPRSdestroyprob, libxprs), Cint, (XPRSprob,), _prob)
 end

@@ -42,7 +42,7 @@ function XPRSgetdaysleft(daysleft)
     ccall((:XPRSgetdaysleft, libxprs), Cint, (Ptr{Cint},), daysleft)
 end

-function XPRSsetcheckedmode(checked_mode)
+function XPRSsetcheckedmode(checked_mode::Cint)
     ccall((:XPRSsetcheckedmode, libxprs), Cint, (Cint,), checked_mode)
 end

@@ -62,1123 +62,1123 @@ function XPRSendlicensing()
     ccall((:XPRSendlicensing, libxprs), Cint, ())
 end

-function XPRSgetlicerrmsg(msg, len)
+function XPRSgetlicerrmsg(msg, len::Cint)
     ccall((:XPRSgetlicerrmsg, libxprs), Cint, (Cstring, Cint), msg, len)
 end

-function XPRSsetlogfile(prob, logname)
+function XPRSsetlogfile(prob::XPRSprob, logname)
     ccall((:XPRSsetlogfile, libxprs), Cint, (XPRSprob, Cstring), prob, logname)
 end

-function XPRSsetintcontrol(prob, _index, _ivalue)
+function XPRSsetintcontrol(prob::XPRSprob, _index::Cint, _ivalue::Cint)
     ccall((:XPRSsetintcontrol, libxprs), Cint, (XPRSprob, Cint, Cint), prob, _index, _ivalue)
 end

-function XPRSsetintcontrol64(prob, _index, _ivalue)
+function XPRSsetintcontrol64(prob::XPRSprob, _index::Cint, _ivalue::Clong)
     ccall((:XPRSsetintcontrol64, libxprs), Cint, (XPRSprob, Cint, Clong), prob, _index, _ivalue)
 end

-function XPRSsetdblcontrol(prob, _index, _dvalue)
+function XPRSsetdblcontrol(prob::XPRSprob, _index::Cint, _dvalue::Cdouble)
     ccall((:XPRSsetdblcontrol, libxprs), Cint, (XPRSprob, Cint, Cdouble), prob, _index, _dvalue)
 end

-function XPRSinterrupt(prob, reason)
+function XPRSinterrupt(prob::XPRSprob, reason::Cint)
     ccall((:XPRSinterrupt, libxprs), Cint, (XPRSprob, Cint), prob, reason)
 end

-function XPRSgetprobname(prob, _svalue)
+function XPRSgetprobname(prob::XPRSprob, _svalue)
     ccall((:XPRSgetprobname, libxprs), Cint, (XPRSprob, Cstring), prob, _svalue)
 end

-function XPRSgetqobj(prob, _icol, _jcol, _dval)
+function XPRSgetqobj(prob::XPRSprob, _icol::Cint, _jcol::Cint, _dval)
     ccall((:XPRSgetqobj, libxprs), Cint, (XPRSprob, Cint, Cint, Ptr{Cdouble}), prob, _icol, _jcol, _dval)
 end

-function XPRSsetprobname(prob, _svalue)
+function XPRSsetprobname(prob::XPRSprob, _svalue)
     ccall((:XPRSsetprobname, libxprs), Cint, (XPRSprob, Cstring), prob, _svalue)
 end
:
```

There's more changes but I think this diff is sufficient for this discussion. Let me know if you want to see the full file, I can commit and revert.

This is strictly typed for some arguments (`Cint` and `Ptr{Nothing}`) but not others (`CString`, `Ptr{Cint}` etc). I think we do need strictly typed functions that users call but it shouldn't be the functions in `src/lib.jl`.

For example, the following code:

```
julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> xp = Xpress.XpressProblem()
Xpress.XpressProblem(Ptr{Nothing} @0x000000013ae14000)

julia> Xpress.getprobname(xp)
ERROR: MethodError: no method matching XPRSgetprobname(::Xpress.XpressProblem, ::Cstring)
Closest candidates are:
  XPRSgetprobname(::Ptr{Nothing}, ::Any) at /Users/USER/GitRepos/Xpress.jl/src/lib.jl:90
Stacktrace:
 [1] invoke(::Function, ::Int64, ::Type{String}, ::Xpress.XpressProblem) at /Users/USER/GitRepos/Xpress.jl/src/utils.jl:21
 [2] getprobname(::Xpress.XpressProblem) at /Users/USER/GitRepos/Xpress.jl/src/api.jl:593
 [3] top-level scope at REPL[5]:1

julia>
```

doesn't work with this strictly typed version of `src/lib.jl`. 

But with the current latest commit on this branch:

```
julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> xp = Xpress.XpressProblem()
Xpress.XpressProblem(Ptr{Nothing} @0x000000012e0b5000)

julia> Xpress.getprobname(xp)
""

julia>
```

it works fine. This is because I've created a [`struct XpressProblem`](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-4cd08f5188a8e8524aa70a7f07808f0aR42-R63):

```
abstract struct CWrapper end
struct XpressProblem <: CWrapper
    ptr::Lib.XPRSprob
end
```

that just holds the `Ptr{Nothing}` to the Xpress problem and I've defined the following `unsafe_convert` on it:

```
Base.unsafe_convert(T::Type{Ptr{Nothing}}, t::CWrapper) = t.ptr
```

`unsafe_convert` is called during `ccall`.

This function (`getprobname`) is strictly typed in `src/api.jl` and only allows users to pass in a type we have defined (`XpressProblem`). We can define additional functions on `XpressProblem`, for example some nicer `show` functionality.

The reason the strictly typed version does not work is that it requires the caller to pass in `Ptr{Nothing}`. Which means we can't use `convert` or `unsafe_convert`. Which also means we will still need to use a `XpressProblem` like `Ptr{Nothing}` holder and pass `p.ptr` explicitly. This means we have to strictly type `src/api.jl` AND `src/lib.jl`.

We will also have to convert do a lot of the manual conversion of types e.g `Int64` to `Cint` etc, which Julia does for us automatically during the `ccall` for us now. And we still have to manage conversions for `Cstring` and `Ptr{Int}` as we are doing now anyway. So I don't see any benefit to making the functions have strict types in `src/lib.jl`.

I completely agree with your point though, that strictly typed is better when interfacing with C code. We don't want segfaults. I think the best way to go about it is to make `src/lib.jl` not strict but make `src/api.jl` as strict as possible. It's more manual work though (which I'm prepared to do). I've done it in [other](https://github.com/dss-extensions/OpenDSSDirect.jl) [projects](https://github.com/GMLC-TDC/HELICS.jl) and I think it works well in providing a nice Julian interface (more intuitive calling for users, more explicit error messages when things go wrong, etc). My only concern with the current approach is that it is a lot of code repetition but I can't see a better way to provide a nice interface.

Thoughts? I'm open to feedback if you think I've missed something.

Reply:
Thanks for all the info.

I agree that Clang's types don't seem very useful. The need for a `CString` workaround and loosing the auto conversion for `Int64` to `Cint` are good examples of the problem.

Code repetition doesn't look that bad, I am afraid of the manual generation part. If it was automatically generated it could work well....

Brainstorming now:

About types...

What types would we need in `api.jl`?
`Cint` <- `Int32` and `Int64`
`Cdouble` <- `Float64`
`Cstring` < -`String` (maybe some `AbstractString` ?)
`Ptr{Cvoid}` <- `Ptr{Nothing}` / `XpressProblem`
`Ptr{Cint}` <- `Vector{Int32}`, `Vector{Int64}` and `Vector{Nothing}` (because we might need it to be C_NULL)
`Ptr{Cdouble}` <- `Vector{Float64}` and `Vector{Nothing}`

Looks doable to automated in `api.jl`?

Reply:
I would say yes it's possible, but not straightforward to cover all ways C code needs to be wrapped.

For example, take `getlicerrmsg`. In `src/lib.jl` its arguments are [`XPRSgetlicerrmsg(::Cstring, ::Cint)`](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-d7b1ad6cb56b02c302a7572a14873d38R65-R67) but in `src/api.jl` it is [`getlicerrmsg()`](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-9134b69774e09f102e883a13c38c0ab9R416-R421).

Another example is functions like `XPRSgetprobname` and `XPRSsetprobname`. Both have type signatures as `Tuple{Ptr{Nothing}, Cstring}`. But `getprobname` takes `xp::XpressProblem` and returns `String` and `setprobname` takes `xp::XpressProblem` and `name::String` and returns `nothing::Nothing`.

I guess we can automate this based on the names (`get`/`set`) etc but I need to look into all the functions to know if there are other nuances.

I think the main thing I'm aiming for now is that in `src/api.jl` we should have as minimal contents in the body of each function as possible. It should be easy to understand for people so that we can easily debug / extend and potentially automate in the future. Currently my plan is to do that using macros. See [getprobname](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-9134b69774e09f102e883a13c38c0ab9R593) as an example.

Reply:
So, first generate Clang's `lib.jl` the post processes that with protected types

Reply:
Ok...

> ( I have another script that gets the docstrings from the xpress documentation website and adds them to the appropriate function in a julia file )

Please add the automation files to the git repo.

> If api.jl is strictly typed for all input arguments, I don't see a benefit to doing this though

The plan is to enforce type with `Union`s, right? so that the user can call with both `Int32` and `Int64`

> I think I'm going to focus on getting api.jl working first so that work on the MOI wrapper can continue

Lets do it! Lets get MOI up and running. You are planning to work on the MOI_wrapper in this same PR, right?

Reply:
Yeah, we are working on both in close coordination